datafusion/core/tests/sql/timestamp.rs

```diff
@@ -1065,3 +1065,79 @@ async fn cast_to_timestamp_micros_twice() -> Result<()> {
 
     Ok(())
 }
+
+#[tokio::test]
+async fn date_bin() {
+    let ctx = SessionContext::new();
+
+    let sql = "SELECT DATE_BIN(INTERVAL '15 minutes', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTAMP '1970-01-01T00:00:00Z') AS res";
+    let results = execute_to_batches(&ctx, sql).await;
+    let expected = vec![
+        "+---------------------+",
+        "| res                 |",
+        "+---------------------+",
+        "| 2022-08-03 14:30:00 |",
+        "+---------------------+",
+    ];
+    assert_batches_eq!(expected, &results);
+
+    // Shift forward by 5 minutes
+    let sql = "SELECT DATE_BIN(INTERVAL '15 minutes', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTAMP '1970-01-01T00:05:00Z') AS res";
+    let results = execute_to_batches(&ctx, sql).await;
+    let expected = vec![
+        "+---------------------+",
+        "| res                 |",
+        "+---------------------+",
+        "| 2022-08-03 14:35:00 |",
+        "+---------------------+",
+    ];
+    assert_batches_eq!(expected, &results);
+
+    // Shift backward by 5 minutes
+    let sql = "SELECT DATE_BIN(INTERVAL '15 minutes', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTAMP '1970-01-01T23:55:00Z') AS res";
+    let results = execute_to_batches(&ctx, sql).await;
+    let expected = vec![
+        "+---------------------+",
+        "| res                 |",
+        "+---------------------+",
+        "| 2022-08-03 14:25:00 |",
+        "+---------------------+",
+    ];
+    assert_batches_eq!(expected, &results);
+
+    // origin after source, timestamp in previous bucket
+    let sql = "SELECT DATE_BIN(INTERVAL '15 minutes', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTAMP '2022-08-03 14:40:00Z') AS res";
+    let results = execute_to_batches(&ctx, sql).await;
+    let expected = vec![
+        "+---------------------+",
+        "| res                 |",
+        "+---------------------+",
+        "| 2022-08-03 14:25:00 |",
+        "+---------------------+",
+    ];
+    assert_batches_eq!(expected, &results);
+
+    // stride by 7 days
+    let sql = "SELECT DATE_BIN(INTERVAL '7 days', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTAMP '1970-01-01 00:00:00Z') AS res";
+    let results = execute_to_batches(&ctx, sql).await;
+    let expected = vec![
+        "+---------------------+",
+        "| res                 |",
+        "+---------------------+",
+        "| 2022-07-28 00:00:00 |",
+        "+---------------------+",
+    ];
+    assert_batches_eq!(expected, &results);
+
+    // origin shifts bins forward 1 day
+    let sql = "SELECT DATE_BIN(INTERVAL '7 days', TIMESTAMP '2022-08-03 14:38:50Z', TIMESTAMP '1970-01-02 00:00:00Z') AS res";
+    let results = execute_to_batches(&ctx, sql).await;
+    let expected = vec![
+        "+---------------------+",
+        "| res                 |",
+        "+---------------------+",
+        "| 2022-07-29 00:00:00 |",
+        "+---------------------+",
+    ];
+    assert_batches_eq!(expected, &results);
+}
```

Comment:
I recommend adding a test showing this function working on actual values (to exercise the `Array` path not just the `ScalarValue`):

Perhaps the examples from https://github.com/apache/arrow-datafusion/issues/3015 👍 

```sql
SELECT 
  DATE_BIN(INTERVAL '15' minute, CAST(time as TIMESTAMP), TIMESTAMP '2001-01-01') AS time, 
  val 
FROM (
  VALUES 
    ('2021-06-10 17:05:00Z', 0.5),
    ('2021-06-10 17:19:10Z', 0.3)
  ) as t (time, val);
```

and 

```sql
SELECT 
  DATE_BIN(INTERVAL '15' minute, CAST(time as TIMESTAMP), TIMESTAMP '2001-01-01 00:02:30') AS time, 
  val 
FROM (
  VALUES 
    ('2021-06-10 17:05:00Z', 0.5),
    ('2021-06-10 17:19:10Z', 0.3)
  ) as t (time, val);
```

```

Reply:
Also, when I tried with `origin` that was a column (rather than a scalar) I got: 

```sql
-- Try using a column value as the origin:
SELECT
  DATE_BIN(INTERVAL '15' minute, CAST(time as TIMESTAMP), CAST(origin as TIMESTAMP)) AS time,
  val
FROM (
  VALUES
    ('2021-06-10 17:05:00Z', '2001-01-01 00:02:30', 0.5),
    ('2021-06-10 17:19:10Z', '2001-01-01 00:02:30', 0.3)
  ) as t (time, origin, val);

ArrowError(ExternalError(Execution("origin of DATE_BIN is an invalid type")))
❯ 
```

(it is probably worth a test for that case too, even if we don't support it in the initial implementation)

Reply:
I'll add the test – and then we can add array support for both the `stride` and `origin` arguments in a future PR.